### PR TITLE
Use `&IndexSlice` instead of `&IndexVec` where possible

### PR DESCRIPTION
**Pull request by scottmcm**
All the same reasons as for `[T]`: more general, less pointer chasing, and `&mut IndexSlice` emphasizes that it doesn't change *length*.

r? @ghost

### CODE REVIEW

**Comment by scottmcm** [comment]
(Update was just a rebase atop #109849, since it touched a couple of the same places.)

File: compiler/rustc_index/src/vec.rs [hunk 1]
```diff
@@ -116,7 +130,7 @@ impl<I: Idx, T> IndexVec<I, T> {
     }
 
     #[inline]
-    pub fn from_elem<S>(elem: T, universe: &IndexVec<I, S>) -> Self
+    pub fn from_elem<S>(elem: T, universe: &IndexSlice<I, S>) -> Self
```
Comment (WaffleLapkin) [inline compiler/rustc_index/src/vec.rs:133]: It would be nice to have documentation here, I don't really get why/how you'd use this method...

**Review by WaffleLapkin — APPROVED** [review]

File: compiler/rustc_index/src/vec.rs (hunk 1 above)
Comment (scottmcm) [reply compiler/rustc_index/src/vec.rs:133]: Send as PR #109913

**Pull request merged**